### ISSUE TITLE
Need another example: "canvas" painting

### ISSUE THREAD

**Issue by Alexey-T** [genericptr/Cocoa-Pascal-Examples#1]
I talked with Dmitry (developer of Cocoa WS) about why Canvas rendering is slow on cocoa. We used the demo to see that it's slow, output time is 140ms on my Mac.
https://github.com/Alexey-T/FreePascal-tests/tree/master/Cocoa%20rendering%20speed%20test

Dmitry asked me to write the same demo using pure Cocoa (not TCanvas). to detect - is this TCanvas fault nor not.
Maybe you can help?

**Comment by genericptr** [comment]
There's too many variables to know why and writing to bitmaps is always slow. What are the speeds like on Windows? I have no idea how text rendering is done in Lazarus but there's lots of ways that could go wrong also as it's transferred down to the cocoa layer.

**Comment by Alexey-T** [comment]
Speed (in the same demo) on Win10 is much much faster - IIRC it's like 20-40ms per frame, while Cocoa is 140-150ms.

**Comment by genericptr** [comment]
In that case it's probably the Cocoa layer in Lazarus. The underlying bitmap context is CoreGraphics which is not likely to perform much different from any other bitmap drawing so it's all going to come down to the abstraction layer in Lazarus.

**Comment by Alexey-T** [comment]
Maybe you can write a demo which does the __similar test__ as my demo does?
this will convict Dmitry that he needs more optimizing.

**Comment by genericptr** [comment]
Maybe if I have time. Are you rendering all the text in the editor using bitmap contexts? You probably want to do this with OpenGL anyways if you want it to be faster. 20-40ms per frame is already really slow, but i guess that's for 1000 passes so I'm not sure how much a full page would take.

**Comment by Alexey-T** [comment]
>Are you rendering all the text in the editor using bitmap contexts? 

I handle TPanel.OnPaint and use Panel's Canvas to paint on it. (I don't use Bitmap but render on panel's canvas).

**Comment by genericptr** [comment]
Yes but how is Cocoa layer handling this? I think it's pushing drawing commands to a bitmap (like `C.TextOut(x, y, 'Colored text string');`)  and then updating something lower level in the system layer.

**Comment by Alexey-T** [comment]
Cocoa layer does this

```
function TCocoaWidgetSet.TextOut(DC: HDC; X,Y: Integer; Str: Pchar; Count: Integer) : Boolean;
begin
  Result := ExtTextOut(DC, X, Y, 0, nil, Str, Count, nil);
end;
```
it calls this
```
function TCocoaWidgetSet.ExtTextOut(DC: HDC; X, Y: Integer; Options: Longint;
  Rect: PRect; Str: PChar; Count: Longint; Dx: PInteger): Boolean;
var
  ctx: TCocoaContext;
begin
  ctx := CheckDC(DC);
  Result := Assigned(ctx);
  if Assigned(ctx) then
    ctx.TextOut(X, Y, Options, Rect, Str, Count, Dx);
end;
```
it calls this
```

procedure TCocoaContext.TextOut(X, Y: Integer; Options: Longint; Rect: PRect; UTF8Chars: PChar; Count: Integer; CharsDelta: PInteger);
var
  BrushSolid, FillBg: Boolean;
  q0, q1: qword;
begin
  q0:= gettickcount64;
  CGContextSaveGState(CGContext());
  q1:= gettickcount64;
  inc(_cocoa_ctx, q1-q0);
  q0:= q1;

  if Assigned(Rect) then
  begin
    // fill background
    //debugln(['TCocoaContext.TextOut ',UTF8Chars,' ',dbgs(Rect^)]);
    if (Options and ETO_OPAQUE) <> 0 then
    begin
      BrushSolid := BkBrush.Solid;
      BkBrush.Solid := True;
      with Rect^ do
        Rectangle(Left, Top, Right, Bottom, True, BkBrush);
      BkBrush.Solid := BrushSolid;
    end;

    if ((Options and ETO_CLIPPED) <> 0) and (Count > 0) then
    begin
      CGContextBeginPath(CGContext);
      CGContextAddRect(CGContext, RectToCGrect(Rect^));
      CGContextClip(CGContext);
    end;
  end;

  q1:= gettickcount64;
  inc(_cocoa_rect, q1-q0);
  q0:= q1;

  if (Count > 0) then
  begin
    FillBg := BkMode = OPAQUE;
    if FillBg then
      FText.BackgroundColor := BkBrush.ColorRef;
    FText.SetText(UTF8Chars, Count);

    q1:= gettickcount64;
    inc(_cocoa_text_set, q1-q0);
    q0:= q1;

    FText.Draw(ctx, X, Y, FillBg, CharsDelta);

    q1:= gettickcount64;
    inc(_cocoa_text_draw, q1-q0);
    q0:= q1;
  end;

  CGContextRestoreGState(CGContext());

  q1:= gettickcount64;
  inc(_cocoa_ctx, q1-q0);
  q0:= q1;

  AttachedBitmap_SetModified();
end;
```

**Comment by genericptr** [comment]
What does `FText.Draw(ctx, X, Y, FillBg, CharsDelta);` call down to? Also `AttachedBitmap_SetModified();` is of interest.

**Comment by Alexey-T** [comment]
FText.Draw is
https://gitlab.com/freepascal.org/lazarus/lazarus/-/blob/main/lcl/interfaces/cocoa/cocoagdiobjects.pas#L1484

Line 1484.

AttachedBitmap_SetModified is empty
https://gitlab.com/freepascal.org/lazarus/lazarus/-/blob/main/lcl/interfaces/cocoa/cocoagdiobjects.pas#L2327

**Comment by genericptr** [comment]
yeah that's something worth profiling. It's also possible it's something as simple as the Cocoa layer forcing a redraw every frame when the Windows layer is flushing after all draw calls are made. When is flushing happen do you know? I expected AttachedBitmap_SetModified to do that.

**Comment by Alexey-T** [comment]
I don't know about flushing. I asked Dmitry to look at this GH thread. @skalogryz

**Comment by genericptr** [comment]
This is for drawing a text editor right? Do you call the entire visible content of the document using a single TextOut call? It's possible that Cocoa text rendering is slow but that's the lowest level you can call down to in Cocoa.

**Comment by Alexey-T** [comment]
>This is for drawing a text editor right?

For drawing the text editor CudaText / and for my cocoa-speed-test. (I gave the link before).

> Do you call the entire visible content of the document using a single TextOut call? 

no! TextOut is called many times! per each string on canvas.

**Comment by Alexey-T** [comment]
E.g. if CudaText opens big C++ file, it will call TextOut count_of_tokens_per_line * count_of_lines, it's about 300-600 times.

**Comment by genericptr** [comment]
"tokens" being a range of styled text? That could be a problem right there that you're not able to make a single draw call per line.

**Comment by Alexey-T** [comment]
Yes, 'token' is simple c++ code fragment (one operator, one Id, one string)

**Comment by Alexey-T** [comment]
So my app CudaText makes lot of small TextOut's for syntax highlighted text. It cannot do it via OpenGL (because I use TCanvas features - bitmaps, Line, TextOut, python API supports Canvas). if Cocoa TextOut is slow the entire CudaText is slow..
can we find the 'slow part of Lazarus Cocoa code'?

**Comment by genericptr** [comment]
I am curious about this myself since recently I was doing text rending with OpenGL so it would be fun to make a test project with Cocoa. It's not totally trivial though so I would need to do this in my spare time.

How are you doing cursors/selection? Does Lazarus keep track of the render frames so you can query it? I know NSLayoutManager does this also so maybe it's calling down into that again.

**Comment by Alexey-T** [comment]
Cursors/selection? In that cocoa-speed-demo we don’t have them. Demo only paints text, and rectangles. Demo has 4 options what  to paint. 

> Does Lazarus keep track of the render frames so you can query it?

No it doesn’t

**Comment by genericptr** [comment]
Ok I had some time to look at this and the problem is that `TCocoaTextLayout.SetText` triggers a recalculation of the layout, which happens every time you call TextOut. For example if if set the string first and then draw 100 lines 100 times it takes 20ms but if I set the string before drawing it takes 200ms.

The solution is to build a single attributed string (which means text + style information) and THEN call TextOut. I don't think the Lazarus control supports this so it needs to be extended.

**Comment by Alexey-T** [comment]
Yes, Lazarus LCL don't support such stages: it needs that we 

- set TCanvas.Brush.Style / Brush.Color 
- set TCanvas.Font.Style / Font.Color
- call TextOut(x, y, text)

so it's bad, not fixable? or we can add 'stages'? @skalogryz

**Comment by genericptr** [comment]
It's probably just better to use FreeType and to render fonts and then draw the characters yourself. NSLayoutManager seems to be designed for larger text ranges and non-mono fonts but if you're updating single characters or short pieces of text it may just be inherently slow.

**Comment by Alexey-T** [comment]
Ryan, Win32 doesn't have any layout-manager for strings, and it works very fast! why?

**Comment by genericptr** [comment]
It may be a lower level API which is better for drawing single strings. If the text is mono space and doesn't wrap NSLayoutManager is probably a bad choice. He could try to reimplement in using https://developer.apple.com/documentation/coretext which may have faster code paths for what you're doing.
